drivers/wifi/nrf700x/zephyr/src/qspi/src/ficr_prog.c

```diff
@@ -0,0 +1,452 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+/* @file
+ * @brief NRF Wi-Fi radio FICR programming functions
+ */
+
+#include <zephyr/logging/log.h>
+#include <zephyr.h>
+#include "rpu_hw_if.h"
+#include "ficr_prog.h"
+
+
+LOG_MODULE_DECLARE(otp_prog, CONFIG_WIFI_LOG_LEVEL);
+
+static void write_word(uint32_t addr, uint32_t data)
+{
+	rpu_write(addr, &data, 4);
+}
+
+static void read_word(uint32_t addr, uint32_t *data)
+{
+	rpu_read(addr, data, 4);
+}
+
+uint32_t check_protection(uint32_t *buff, uint32_t off1, uint32_t off2,
+		    uint32_t off3, uint32_t off4)
+{
+	if ((buff[off1] == 0x00000000) &&
+		(buff[off2] == 0x00000000)  &&
+		(buff[off3] == 0x00000000) &&
+		(buff[off4] == 0x00000000))
+		return 0x00000000;
+	else if ((buff[off1] == 0xFFFFFFFF) &&
+		(buff[off2] == 0xFFFFFFFF) &&
+		(buff[off3] == 0xFFFFFFFF) &&
+		(buff[off4] == 0xFFFFFFFF))
+		return 0xFFFFFFFF;
+	else if ((buff[off1] == 0x50FA50FA) &&
+		(buff[off2] == 0x50FA50FA) &&
+		(buff[off3] == 0x50FA50FA) &&
+		(buff[off4] == 0x50FA50FA))
+		return 0x50FA50FA;
+	else
+		return 0xdeadbeef;
+}
+
+
+static void set_otp_timing_reg_40mhz(void)
+{
+	uint32_t val;
+
+	val = OTP_TIMING_REG1_VAL;
+	write_word(OTP_TIMING_REG1_ADDR, val);
+
+	val = OTP_TIMING_REG2_VAL;
+	write_word(OTP_TIMING_REG2_ADDR, val);
+}
+
+static int poll_otp_ready(void)
+{
+
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while ((otp_mem_status != OTP_READY) && (poll != 100)) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x4) == OTP_READY) {
+			return 0;
+		}
+		poll += 1;
+	}
+
+	LOG_INF("OTP is not ready\n");
+	return -ENOEXEC;
+}
+
+
+static int req_otp_standby_mode(void)
+{
+
+	uint32_t val;
+	int32_t otp_ready_exception;
+
+	val = 0x0;
+	write_word(OTP_RWSBMODE_ADDR, val);
+	otp_ready_exception = poll_otp_ready();
+	return otp_ready_exception;
+}
+
+
+static int otp_wr_voltage_2V5(void)
+{
+	uint32_t val = OTP_VOLTCTRL_2V5;
+	int32_t err;
+
+	err = req_otp_standby_mode();
+
+	if (err) {
+		LOG_INF("Failed Setting OTP voltage IOVDD to 2.5V\n");
+		return -ENOEXEC;
+	}
+	write_word(OTP_VOLTCTRL_ADDR, val);
+	return 0;
+}
+
+static int poll_otp_read_valid(void)
+{
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while ((otp_mem_status != OTP_READ_VALID) && (poll < 100)) {
+		read_word(OTP_POLL_ADDR, &otp_mem_status);
+
+		if ((otp_mem_status&0x2) == OTP_READ_VALID) {
+			return 0;
+		}
+		poll += 1;
+	}
+	LOG_INF("failed poll_otp_read_valid()\n");
+	return -ENOEXEC;
+}
+
+static int poll_otp_wrdone(void)
+{
+	int otp_mem_status = 0;
+	int poll = 0;
+
+	while ((otp_mem_status != OTP_WR_DONE) && (poll < 100)) {
```

Comment:
Should this also be checked against (otp_mem_status & 0x1) as is being done below?

Reply:
Done